### PR TITLE
feat: cuda implementation for `ggml_conv_transpose_1d`

### PR DESCRIPTION
**Pull request by balisujohn**
adds a cuda implementation for `ggml_conv_transpose_1d`

may extend the CPU version of the op to allow padding (currently, padding must be zero). 

Currently a draft, posting now to prevent duplication of effort, I will remove draft status when ready for review.

### CODE REVIEW

**Comment by balisujohn** [comment]
Alright, this is  ready for review. The key questions/concerns I have are as follows:

I added some tests for `ggml_conv_transpose_1d` in `test-conv-transpose-1d.cpp` some of these tests mirror the conv1 1d transpose tests in `test-conv-transpose.c` I have verified that the tests in `test-conv-transpose-1d.cpp` pass in both cuda and CPU mode, but I'm not sure if what's currently in the PR will run the tests in CUDA mode in the CI/CD, I had to remove the manual ```#define GGML_USE_CUBLAS```I had added in order for the CI/CD build to work. Curious how best to combine the new tests I added with what's already there, and how to make sure the tests I added are actually running with CUDA in the CI/CD process.

There is an extremely bizarre problem I'm encountering when trying to use this op with tortoise, the result is all zeros, but if this line is changed to 
````
    dst[global_index] = accumulator;
````
````
    dst[global_index] += accumulator;
````
, the result is no longer all zeros.
I'm going to try to isolate the error into a self-contained example in a ggml fork so other people can take a look at it.

**Comment by slaren** [comment]
The github runners cannot run CUDA code. There is the ggml-ci can run CUDA tests, but it only monitors branches in this repository. A test should be added to `test-backend-ops` as well.

File: src/ggml-cuda/conv-transpose-1d.cu [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        const int dst_ne0, const int dst_ne1, const int dst_ne2, const int dst_ne3,
+        const float * src0, const float * src1,  float * dst) {
+    int global_index = threadIdx.x + blockIdx.x * blockDim.x;
+    if (global_index >= output_size) {
+        return;
+    }
+
+    int out_index = global_index / dst_ne0;
+
+    int accumulator = 0;
```
[the remark this answers is not in the document]
Comment (balisujohn) [reply src/ggml-cuda/conv-transpose-1d.cu:16]: yeah that seems to fix the issue I was experiencing; it's bizarre that the tests still passed even in cuda mode with the types accidentally set to `int` Thanks so much!

**Comment by balisujohn** [comment]
I fixed the accumulator bug, added a test to test-backend-ops; ready for review again :^)

File: src/ggml-cuda/conv-transpose-1d.cu [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    const float * src0_d = (const float *)src0->data;
+
+    const ggml_tensor * src1 = dst->src[1];
+    const float * src1_d = (const float *)src1->data;
+
+    float * dst_d = (float *)dst->data;
+    cudaStream_t stream = ctx.stream();
+
+    GGML_ASSERT(src0->type == GGML_TYPE_F32);
+    GGML_ASSERT( dst->type == GGML_TYPE_F32);
```
Comment (ggerganov) [inline src/ggml-cuda/conv-transpose-1d.cu:73]: Also assert contiguous `src0` and `src1`

**Review by ggerganov — COMMENTED** [review]
I don't see update of `ggml_backend_cuda_supports_op()` to indicate that this operation is now supported. While at it, update the rest of the GPU backends to return `false` in their corresponding `supports_op()` functions so that the tests do not fail

Maybe add a few more small tests to `tests-backend-ops` just in case

**Comment by balisujohn** [comment]
I added some more tests to `test-backend-ops` and added the asserts. Shouldn't the other backends already return false in the default case if the OP is not listed?

**Review by slaren — APPROVED** [review]

File: tests/test-backend-ops.cpp [hunk 3]
```diff
@@ -1216,6 +1216,38 @@ struct test_pool2d : public test_case {
...
+    ggml_tensor * build_graph(ggml_context * ctx) override {
+        ggml_tensor * input = ggml_new_tensor(ctx, GGML_TYPE_F32, 4, ne_input.data());
+        ggml_tensor * kernel = ggml_new_tensor(ctx, GGML_TYPE_F32, 4, ne_kernel.data());
+        ggml_tensor * out = ggml_conv_transpose_1d(ctx, kernel, input, s0, p0, d0);
+        return out;
+    }
+};
+
+
+
```
[the remark this answers is not in the document]
Comment (balisujohn) [reply tests/test-backend-ops.cpp]: done

**Pull request merged**